node/pkg/watchers/ibc/watcher.go

```diff
@@ -0,0 +1,734 @@
+		w.logger.Error("failed to query IBC connectionID for channel", zap.String("ibcChannel", evt.ChannelID), zap.Error(err))
+		connectionErrors.WithLabelValues("unexpected_ibc_channel_error").Inc()
+		return
+	}
+
+	ce, exists := w.connectionMap[connectionID]
+	if !exists {
+		w.logger.Error("ignoring an event from an unexpected IBC connection", zap.String("ibcConnection", connectionID))
+		connectionErrors.WithLabelValues("unexpected_ibc_channel_error").Inc()
+		return
+	}
+
+	if evt.Msg.EmitterChain != ce.chainID {
+		w.logger.Error(fmt.Sprintf("chain id mismatch in %s message", observationType),
+			zap.String("ibcConnectionID", connectionID),
+			zap.String("ibcChannelID", evt.ChannelID),
+			zap.Uint16("expectedChainID", uint16(ce.chainID)),
+			zap.Uint16("actualChainID", uint16(evt.Msg.EmitterChain)),
+			zap.Stringer("txHash", txHash),
+			zap.String("msgId", evt.Msg.MessageIDString()),
+		)
+		invalidChainIdMismatches.WithLabelValues(evt.ChannelID).Inc()
+		return
+	}
+
+	w.logger.Info(fmt.Sprintf("%s message detected", observationType),
+		zap.String("ChannelID", evt.ChannelID),
+		zap.String("ConnectionID", ce.connectionID),
+		zap.String("ChainName", ce.chainName),
+		zap.Stringer("TxHash", evt.Msg.TxHash),
+		zap.Stringer("EmitterChain", evt.Msg.EmitterChain),
+		zap.Stringer("EmitterAddress", evt.Msg.EmitterAddress),
+		zap.Uint64("Sequence", evt.Msg.Sequence),
+		zap.Uint32("Nonce", evt.Msg.Nonce),
+		zap.Stringer("Timestamp", evt.Msg.Timestamp),
+		zap.Uint8("ConsistencyLevel", evt.Msg.ConsistencyLevel),
+	)
+
+	ce.msgC <- evt.Msg
+	messagesConfirmed.WithLabelValues(ce.chainName).Inc()
+}
+
+/*
+This query:
+  `{"all_chain_connections":{}}` is `eyJhbGxfY2hhaW5fY29ubmVjdGlvbnMiOnt9fQ==`
+  which becomes:
+  http://localhost:1319/cosmwasm/wasm/v1/contract/wormhole1nc5tatafv6eyq7llkr2gv50ff9e22mnf70qgjlv737ktmt4eswrq0kdhcj/smart/eyJhbGxfY2hhaW5fY29ubmVjdGlvbnMiOnt9fQ%3D%3D
+
+Returns something like this:
+{
+	"data": {
+		"chain_connections": [
+			[
+				"Y29ubmVjdGlvbi0w",
+				18
+			],
+			[
+				"Y29ubmVjdGlvbi00Mg==",
+				22
+			]
+		]
+	}
+}
+
+*/
+
+type ibcAllChainConnectionsQueryResults struct {
+	Data struct {
+		ChainConnections [][]interface{} `json:"chain_connections"`
+	}
+}
+
+var connectionIdMapQuery = url.QueryEscape(base64.StdEncoding.EncodeToString([]byte(`{"all_chain_connections":{}}`)))
+
+// queryConnectionIdMap queries the contract for the set of chain IDs available on IBC and their corresponding IBC connections.
+func (w *Watcher) queryConnectionIdMap() (map[vaa.ChainID]string, error) {
+	connIdMap := make(map[vaa.ChainID]string)
+
+	client := &http.Client{
+		Timeout: time.Second * 5,
+	}
+
+	query := fmt.Sprintf(`%s/cosmwasm/wasm/v1/contract/%s/smart/%s`, w.lcdUrl, w.contractAddress, connectionIdMapQuery)
+	connResp, err := client.Get(query)
+	if err != nil {
+		w.logger.Error("channel map query failed", zap.String("query", query), zap.Error(err))
+		return connIdMap, fmt.Errorf("query failed: %w", err)
```

Comment:
best to return nil to make sure any attempted reads panic when an error is returned

Reply:
Done